populator-machinery/controller.go

```diff
@@ -559,101 +692,104 @@ func (c *controller) syncPvc(ctx context.Context, key, pvcNamespace, pvcName str
+				// Make the pod
+				pod = &corev1.Pod{
 					ObjectMeta: metav1.ObjectMeta{
-						Name:      pvcPrimeName,
+						Name:      podName,
 						Namespace: c.populatorNamespace,
 					},
-					Spec: corev1.PersistentVolumeClaimSpec{
-						AccessModes:      pvc.Spec.AccessModes,
-						Resources:        pvc.Spec.Resources,
-						StorageClassName: pvc.Spec.StorageClassName,
-						VolumeMode:       pvc.Spec.VolumeMode,
-					},
+					Spec: makePopulatePodSpec(pvcPrimeName),
 				}
-				if waitForFirstConsumer {
-					pvcPrime.Annotations = map[string]string{
-						annSelectedNode: nodeName,
+				pod.Spec.Volumes[0].VolumeSource.PersistentVolumeClaim.ClaimName = pvcPrimeName
+				con := &pod.Spec.Containers[0]
+				con.Image = c.podConfig.ImageName
+				con.Args = args
+				if rawBlock {
+					con.VolumeDevices = []corev1.VolumeDevice{
+						{
+							Name:       populatorPodVolumeName,
+							DevicePath: c.podConfig.DevicePath,
+						},
+					}
+				} else {
+					con.VolumeMounts = []corev1.VolumeMount{
+						{
+							Name:      populatorPodVolumeName,
+							MountPath: c.podConfig.MountPath,
+						},
 					}
 				}
-				_, err = c.kubeClient.CoreV1().PersistentVolumeClaims(c.populatorNamespace).Create(ctx, pvcPrime, metav1.CreateOptions{})
+				if waitForFirstConsumer {
+					pod.Spec.NodeName = nodeName
+				}
+				_, err = c.kubeClient.CoreV1().Pods(c.populatorNamespace).Create(ctx, pod, metav1.CreateOptions{})
 				if err != nil {
-					c.recorder.Eventf(pvc, corev1.EventTypeWarning, reasonPVCCreationError, "Failed to create populator PVC: %s", err)
+					c.recorder.Eventf(pvc, corev1.EventTypeWarning, reasonPodCreationError, "Failed to create populator pod: %s", err)
 					return err
 				}
-			}
+				c.recorder.Eventf(pvc, corev1.EventTypeNormal, reasonPodCreationSuccess, "Populator started")
 
-			// We'll get called again later when the pod exists
-			return nil
+				// We'll get called again later when the pod exists
+				return nil
+			}
 		}
 
-		if corev1.PodSucceeded != pod.Status.Phase {
-			if corev1.PodFailed == pod.Status.Phase {
-				c.recorder.Eventf(pvc, corev1.EventTypeWarning, reasonPodFailed, "Populator failed: %s", pod.Status.Message)
-				// Delete failed pods so we can try again
-				err = c.kubeClient.CoreV1().Pods(c.populatorNamespace).Delete(ctx, pod.Name, metav1.DeleteOptions{})
-				if err != nil {
-					return err
+		// If use provider specific implementation, invoke the populateCompleteFn() function. Otherwise, proceed with populator pod
+		if c.providerFunctionConfig.PopulateCompleteFn != nil {
+			complete, err := c.providerFunctionConfig.PopulateCompleteFn(ctx, *params)
+			if err != nil {
+				c.recorder.Eventf(pvc, corev1.EventTypeWarning, reasonPopulateOperationFailed, "Population operation failed: %s", err.Error())
+				return err
+			}
+			if !complete {
+				// Return error to force reque pvc. We'll get called again later when the population operation complete
```

Comment:
Can be a TODO for later, but we may need to revisit this when thinking about metrics. We may not want to record this as an error.

Reply:
Sure, we can revisit the design. This[1] should help skip log&metrics

[1]https://github.com/kubernetes-csi/lib-volume-populator/pull/133/files#diff-d56d5d91bc0d728e3a4b0c81c1d5c2188dd3a1bb53cfe3a01fa25f724f29ead6R517